src/DDTrace/OpenTelemetry/SpanContext.php

```diff
@@ -0,0 +1,130 @@
+
+final class SpanContext implements SpanContextInterface
+{
+    /** @var SpanData */
+    private $span;
+
+    private bool $sampled;
+
+    private bool $remote;
+
+    private ?string $traceId;
+
+    private ?string $spanId;
+
+    private bool $isValid = true;
+
+    private function __construct(SpanData $span, bool $sampled, bool $remote, ?string $traceId = null, ?string $spanId = null)
+    {
+        $this->span = $span;
+        $this->sampled = $sampled;
+        $this->remote = $remote;
+        $this->traceId = $traceId ?: \DDTrace\root_span()->traceId;
+        $this->spanId = $spanId ?: $this->span->hexId();
+
+        // TraceId must be exactly 16 bytes (32 chars) and at least one non-zero byte
+        // SpanId must be exactly 8 bytes (16 chars) and at least one non-zero byte
+        if (!SpanContextValidator::isValidTraceId($this->traceId) || !SpanContextValidator::isValidSpanId($this->spanId)) {
+            $this->traceId = SpanContextValidator::INVALID_TRACE;
+            $this->spanId = SpanContextValidator::INVALID_SPAN;
+            $this->isValid = false;
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getTraceId(): string
+    {
+        return $this->traceId;
+    }
+
+    public function getTraceIdBinary(): string
+    {
+        return hex2bin($this->getTraceId());
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getSpanId(): string
+    {
+        return $this->spanId;
+    }
+
+    public function getSpanIdBinary(): string
+    {
+        return hex2bin($this->getSpanId());
+    }
+
+    public function getTraceState(): ?TraceStateInterface
+    {
+        $traceContext = generate_distributed_tracing_headers(['tracecontext']);
+        return new TraceState($traceContext['tracestate'] ?? null);
+    }
+
+    public function isSampled(): bool
+    {
+        return $this->sampled;
+    }
+
+    public function isValid(): bool
+    {
+        return $this->isValid;
+    }
+
+    public function isRemote(): bool
+    {
+        return $this->remote;
+    }
+
+    public function getTraceFlags(): int
+    {
+        return $this->sampled ? TraceFlags::SAMPLED : TraceFlags::DEFAULT;
+    }
+
+    /** @inheritDoc */
+    public static function createFromRemoteParent(string $traceId, string $spanId, int $traceFlags = TraceFlags::DEFAULT, ?TraceStateInterface $traceState = null): SpanContextInterface
+    {
+        return API\SpanContext::createFromRemoteParent($traceId, $spanId, $traceFlags, $traceState);
+    }
+
+    /** @inheritDoc */
+    public static function create(string $traceId, string $spanId, int $traceFlags = TraceFlags::DEFAULT, ?TraceStateInterface $traceState = null): SpanContextInterface
+    {
+        return API\SpanContext::create($traceId, $spanId, $traceFlags, $traceState);
+    }
+
+    /** @inheritDoc */
+    public static function getInvalid(): SpanContextInterface
+    {
+        return API\SpanContext::getInvalid();
+    }
```

Comment:
These methods are not used at all, right?

Overall, it seems like this class is just decorating OpenTelemetry\API\Trace\SpanContext ... with the only goal of on-the-fly generation of the TraceState object?

Wouldn't it be more straightforward to Implement the TraceStateInterface instead and inject that into the API\SpanContext?

Reply:
Let's say we do a `DDTrace\TraceState` class. Wouldn't it also need to generate trace states on the fly? Isn't that counter-intuitive for a class that should represent an immutable list of key/value pairs? I tried to get my head around this today but I feel like I misunderstood you here 😬 

Imo, we need to generate trace states on the fly since distributed attributes can be added at any point in a span's life, hence changing the trace state 🤔

Reply:
> Wouldn't it also need to generate trace states on the fly

Yes, it would. But, uh, that's sort of "how it is". Like, SpanContext is also supposed to return an immutable tracestate basically, from whatever angle you approach it :-)